pkg/policy/directory/watcher.go

```diff
@@ -0,0 +1,222 @@
+	cnp, err := p.translateToCNPObject(cnpFilePath)
+	if err != nil {
+		return err
+	}
+
+	fileName := filepath.Base(cnpFilePath)
+
+	resourceID := ipcacheTypes.NewResourceID(
+		ipcacheTypes.ResourceKindFile,
+		p.config.StaticCNPPath,
+		fileName,
+	)
+
+	// convert to rules
+	rules, err := cnp.Parse()
+	if err != nil {
+		return err
+	}
+
+	// update labels
+	lbls := getLabels(fileName, cnp)
+	for _, r := range rules {
+		r.Labels = lbls
+	}
+
+	// add to policy engine
+	_, err = p.policyManager.PolicyAdd(rules, &policy.AddOptions{
+		ReplaceByResource: true,
+		Source:            source.Directory,
+		Resource:          resourceID,
+	})
+
+	if err == nil {
+		p.fileNameToCnpCache[fileName] = cnp
+	}
+
+	return err
+}
+
+func (p *policyWatcher) deleteFromPolicyEngine(cnpFilePath string) error {
+	fileName := filepath.Base(cnpFilePath)
+	cnp := p.fileNameToCnpCache[fileName]
+	if cnp == nil {
+		p.log.WithField("file", fileName).Error("BUG: Policy deletion request for file which was never added")
+		return nil
+	}
+
+	resourceID := ipcacheTypes.NewResourceID(
+		ipcacheTypes.ResourceKindFile,
+		p.config.StaticCNPPath,
+		fileName,
+	)
+	_, err := p.policyManager.PolicyDelete(getLabels(fileName, cnp), &policy.DeleteOptions{
+		Source:           source.Directory,
+		DeleteByResource: true,
+		Resource:         resourceID})
+
+	delete(p.fileNameToCnpCache, fileName)
+	return err
+}
+
+func (p *policyWatcher) watchDirectory(ctx context.Context) {
+	go func() {
+		p.log.Info("Directory policy watcher started")
+		watcher, err := fsnotify.NewWatcher()
+		if err != nil {
+			p.log.WithError(err).Error("Initializing NewWatcher failed")
+			return
+		}
+		defer watcher.Close()
+
+		dir := p.config.StaticCNPPath
+		err = watcher.Add(dir)
+		if err != nil {
+			p.log.WithError(err).WithField(logfields.Path, dir).Error("Failed to watch policy directory. Policies will not be loaded from disk")
+			return
+		}
+
+		files, err := os.ReadDir(dir)
+		if err != nil {
+			p.log.WithError(err).WithField("Directory", dir).Warn("Failed to read policy directory")
+			return
+		}
+
+		// Read existing cnp files before watcher was added
+		for _, f := range files {
+			absPath := filepath.Join(dir, f.Name())
+			err := p.addToPolicyEngine(absPath)
+			if err != nil {
+				p.log.WithError(err).WithField(logfields.Path, absPath).Error("Failed to add network policy to policy engine")
+			}
+			reportCNPChangeMetrics(err)
+		}
+		close(p.readStatus)
+		// Listen for file add, update, rename and delete
+		for {
+			select {
+			case event := <-watcher.Events:
+				if filepath.Ext(event.Name) != ".yaml" {
+					continue
+				}
+				if !utf8.ValidString(event.Name) {
+					p.log.WithField(logfields.Path, event.Name).Error("filename has non utf characters")
+					continue
+				}
+				if validation.IsDNS1123Subdomain(filepath.Base(event.Name)) != nil {
+					p.log.WithField("name", filepath.Base(event.Name)).Error("CNP name parse validation failed")
```

Comment:
nit: Let's give the user some context why it failed

```suggestion
				if reasons := validation.IsDNS1123Subdomain(filepath.Base(event.Name)); len(reasons) > 0 {
					p.log.WithFields(logrus.Fields{
						"name", filepath.Base(event.Name),
						"reasons", reasons,
					}).Error("CNP name parse validation failed")
```